site/source/docs/optimizing/Module-Splitting.rst

```diff
@@ -0,0 +1,367 @@
+.. _Module-Splitting:
+
+================
+Module Splitting
+================
+
+*wasm-split and the SPLIT_MODULE Emscripten integration are both in active
+development and may change and gain new features frequently. This page will be
+kept up-to-date with the latest changes.*
+
+wasm-split is a Binaryen tool that splits a Wasm module into a primary and a
+secondary module. The primary module has all the same imports and exports as the
+original module and is meant to be a drop-in replacement for it. However, it
+also imports a placeholder function for each secondary function that was split
+out into the secondary module. Before the secondary module is loaded, calls of
+secondary functions will call the appropriate placeholder function instead. The
+placeholder functions are responsible for loading and instantiating the
+secondary module, which automatically replaces all the placeholder functions
+with the original secondary functions when it is instantiated. After the
+secondary module is loaded, the placeholder function that loaded it is also
+responsible for calling its corresponding newly-loaded secondary function and
+returning the result to its caller. The loading of the secondary module is
+therefore completely transparent to the primary module; it just looks like a
+function call took a long time to return.
```

Comment:
I feel like the first paragraph here describes how wasm-split works at a low and technical level (what happens to imports, placeholders, etc.) but not what it does at a high level. Perhaps it would be good to begin with something like this?

"Often in a large codebase there is a lot of code that is very rarely used in practice. It would be good to separate that code out and only load it on demand, so that startup is faster. One way to do that is to use dynamic linking, however, that has some downsides like it requires the developer to manually refactor the application into libraries, and also dynamic linking requires relocations which add overhead. wasm-split is another approach in which we reduce the initial amount of code downloaded by splitting the wasm after it is built normally. We split the wasm into a *primary* and a *secondary* module. The idea is that the primary module is loaded first and contains the code that is expected to be run, while the secondary module contains code that may never be needed. wasm-split will add code in the primary module that will automatically load the secondary module if it is actually needed."

Reply:
Good idea 👍